--- /usr/include/Vc/global.h	2013-03-26 10:45:14.000000000 +0100
+++ /tmp/global.h	2013-09-03 16:16:51.000000000 +0200
@@ -17,6 +17,8 @@
 
 */
 
+#include<stdio.h>
+
 #ifndef VC_GLOBAL_H
 #define VC_GLOBAL_H
 
@@ -97,6 +99,107 @@
 #define IMPL_MASK 0xFFF00000
 #define EXT_MASK  0x000FFFFF
 
+#if defined(__AVX__)
+#  define VC_IMPL_MASK_AVX AVX
+#endif
+#if defined(__SSE4_2__)
+#  define VC_IMPL_MASK_SSE SSE
+#  define VC_IMPL_MASK_SSE4_2 SSE4_2
+#endif
+#if defined(__SSE4_1__)
+#  define VC_IMPL_MASK_SSE SSE
+#  define VC_IMPL_MASK_SSE4_1 SSE4_1
+#endif
+#if defined(__SSE3__)
+#  define VC_IMPL_MASK_SSE SSE
+#  define VC_IMPL_MASK_SSE3 SSE3
+#endif
+#if defined(__SSSE3__)
+#  define VC_IMPL_MASK_SSE SSE
+#  define VC_IMPL_MASK_SSSE3 SSSE3
+#endif
+#if defined(__SSE2__)
+#  define VC_IMPL_MASK_SSE SSE
+#  define VC_IMPL_MASK_SSE2 SSE2
+#endif
+#ifdef __FMA4__
+#  define VC_IMPL_MASK_FMA4 FMA4
+#endif
+#ifdef __XOP__
+#  define VC_IMPL_MASK_XOP XOP
+#endif
+#ifdef __F16C__
+#  define VC_IMPL_MASK_F16C F16C
+#endif
+#ifdef __POPCNT__
+#  define VC_IMPL_MASK_POPCNT POPCNT
+#endif
+#ifdef __SSE4A__
+#  define VC_IMPL_MASK_SSE4a SSE4a
+#endif
+#ifdef __FMA__
+#  define VC_IMPL_MASK_FMA FMA
+#endif
+
+#ifndef VC_IMPL_MASK_AVX
+#  define VC_IMPL_MASK_AVX 0x00000000
+#endif
+#ifndef VC_IMPL_MASK_SSE4_2
+#  define VC_IMPL_MASK_SSE4_2 0x00000000
+#endif
+#ifndef VC_IMPL_MASK_SSE4_1
+#  define VC_IMPL_MASK_SSE4_1 0x00000000
+#endif
+#ifndef VC_IMPL_MASK_SSE3
+#  define VC_IMPL_MASK_SSE3 0x00000000
+#endif
+#ifndef VC_IMPL_MASK_SSE2
+#  define VC_IMPL_MASK_SSE2 0x00000000
+#endif
+#ifndef VC_IMPL_MASK_FMA4
+#  define VC_IMPL_MASK_FMA4 0x00000000
+#endif
+#ifndef VC_IMPL_MASK_XOP
+#  define VC_IMPL_MASK_XOP 0x00000000
+#endif
+#ifndef VC_IMPL_MASK_F16C
+#  define VC_IMPL_MASK_F16C 0x00000000
+#endif
+#ifndef VC_IMPL_MASK_POPCNT
+#  define VC_IMPL_MASK_POPCNT 0x00000000
+#endif
+#ifndef VC_IMPL_MASK_SSE4a
+#  define VC_IMPL_MASK_SSE4a 0x00000000
+#endif
+#ifndef VC_IMPL_MASK_FMA
+#  define VC_IMPL_MASK_FMA 0x00000000
+#endif
+
+#define XSTR(s) STR(s)
+#define STR(s) s
+
+#define VC_IMPL_MASK ((VC_IMPL_MASK_AVX)    | \
+                      (VC_IMPL_MASK_SSE4_2) | \
+                      (VC_IMPL_MASK_SSE4_1) | \
+                      (VC_IMPL_MASK_SSE3)   | \
+                      (VC_IMPL_MASK_SSE2)   | \
+                      (VC_IMPL_MASK_FMA4)   | \
+                      (VC_IMPL_MASK_XOP)    | \
+                      (VC_IMPL_MASK_F16C)   | \
+                      (VC_IMPL_MASK_POPCNT) | \
+                      (VC_IMPL_MASK_SSE4a)  | \
+                      (VC_IMPL_MASK_FMA))
+
+#ifndef VC_IMPL
+inline void VC_IMPL_MASK_DEBUG(){
+  printf("VC_IMPL_MASK=%8.8x, VC_IMPL is not set\n",VC_IMPL_MASK);
+}
+#else
+inline void VC_IMPL_MASK_DEBUG(){
+  printf("VC_IMPL_MASK=%8.8x, VC_IMPL=%8.8x\n",VC_IMPL_MASK,VC_IMPL);
+}
+#endif
+
 #ifdef VC_MSVC
 # ifdef _M_IX86_FP
 #  if _M_IX86_FP >= 1
@@ -228,22 +331,22 @@
 #    define VC_IMPL_SSE2 1
 #    define VC_IMPL_SSE 1
 #  endif
-#  if (VC_IMPL & XOP)
+#  if (VC_IMPL & XOP & VC_IMPL_MASK)
 #    define VC_IMPL_XOP 1
 #  endif
-#  if (VC_IMPL & FMA4)
+#  if (VC_IMPL & FMA4 & VC_IMPL_MASK)
 #    define VC_IMPL_FMA4 1
 #  endif
-#  if (VC_IMPL & F16C)
+#  if (VC_IMPL & F16C & VC_IMPL_MASK)
 #    define VC_IMPL_F16C 1
 #  endif
-#  if (VC_IMPL & POPCNT)
+#  if (VC_IMPL & POPCNT & VC_IMPL_MASK)
 #    define VC_IMPL_POPCNT 1
 #  endif
-#  if (VC_IMPL & SSE4a)
+#  if (VC_IMPL & SSE4a & VC_IMPL_MASK)
 #    define VC_IMPL_SSE4a 1
 #  endif
-#  if (VC_IMPL & FMA)
+#  if (VC_IMPL & FMA & VC_IMPL_MASK)
 #    define VC_IMPL_FMA 1
 #  endif
 #  undef VC_IMPL
