diff -urP /var/tmp/portage/sci-libs/Vc-0.7.1-r2/work/Vc-0.7.1/cmake/DisableVectorUnits.cmake /tmp/Vc-0.7.1/cmake/DisableVectorUnits.cmake
--- /var/tmp/portage/sci-libs/Vc-0.7.1-r2/work/Vc-0.7.1/cmake/DisableVectorUnits.cmake	2014-02-20 19:52:58.000000000 +0100
+++ /tmp/Vc-0.7.1/cmake/DisableVectorUnits.cmake	2014-02-20 19:52:31.000000000 +0100
@@ -40,9 +40,12 @@
   SET(_SSE_VERSION "00")
   STRING(REPLACE "+" ";" __UNITS "${_UNITS}")
   LIST(APPEND ___UNITS ${__UNITS})
-#  FOREACH(_U IN LISTS ___UNITS)
-#    MESSAGE(STATUS ${_U})
-#  ENDFOREACH()
+  FOREACH(_U IN LISTS ___UNITS)
+    MESSAGE(STATUS ${_U})
+  ENDFOREACH()
+  FOREACH(_U IN LISTS _available_vector_units_list)
+    MESSAGE(STATUS ${_U})
+  ENDFOREACH()
   _my_find(___UNITS "SSE4_2" _found)
   if(_found)
     MESSAGE(STATUS "SSE4_2 found.")
diff -urP /var/tmp/portage/sci-libs/Vc-0.7.1-r2/work/Vc-0.7.1/cmake/OptimizeForArchitecture.cmake /tmp/Vc-0.7.1/cmake/OptimizeForArchitecture.cmake
--- /var/tmp/portage/sci-libs/Vc-0.7.1-r2/work/Vc-0.7.1/cmake/OptimizeForArchitecture.cmake	2014-02-20 19:52:57.000000000 +0100
+++ /tmp/Vc-0.7.1/cmake/OptimizeForArchitecture.cmake	2014-02-20 19:52:31.000000000 +0100
@@ -230,7 +230,7 @@
    elseif(TARGET_ARCHITECTURE STREQUAL "barcelona")
       list(APPEND _march_flag_list "barcelona")
       list(APPEND _march_flag_list "core2")
-      list(APPEND _available_vector_units_list "sse" "sse2" "sse3" "sse4a")
+      list(APPEND _available_vector_units_list "sse" "sse2" "sse3" "sse4a" "popcnt")
    elseif(TARGET_ARCHITECTURE STREQUAL "istanbul")
       list(APPEND _march_flag_list "barcelona")
       list(APPEND _march_flag_list "core2")
@@ -256,6 +256,7 @@
       _my_find(_available_vector_units_list "sse4.1" SSE4_1_FOUND)
       _my_find(_available_vector_units_list "sse4.2" SSE4_2_FOUND)
       _my_find(_available_vector_units_list "sse4a" SSE4a_FOUND)
+      _my_find(_available_vector_units_list "popcnt" POPCNT_FOUND)
       if(DEFINED Vc_AVX_INTRINSICS_BROKEN AND Vc_AVX_INTRINSICS_BROKEN)
          UserWarning("AVX disabled per default because of old/broken compiler")
          set(AVX_FOUND false)
@@ -288,7 +289,8 @@
       set(USE_XOP    ${XOP_FOUND}    CACHE BOOL "Use XOP." ${_force})
       set(USE_FMA    ${FMA_FOUND}    CACHE BOOL "Use FMA." ${_force})
       set(USE_FMA4   ${FMA4_FOUND}   CACHE BOOL "Use FMA4." ${_force})
-      mark_as_advanced(USE_SSE2 USE_SSE3 USE_SSSE3 USE_SSE4_1 USE_SSE4_2 USE_SSE4a USE_AVX USE_XOP USE_FMA USE_FMA4)
+      set(USE_POPCNT ${POPCNT_FOUND} CACHE BOOL "Use POPCNT." ${_force})
+      mark_as_advanced(USE_SSE2 USE_SSE3 USE_SSSE3 USE_SSE4_1 USE_SSE4_2 USE_SSE4a USE_AVX USE_XOP USE_FMA USE_FMA4 USE_POPCNT)
       if(USE_SSE2)
          list(APPEND _enable_vector_unit_list "sse2")
       else(USE_SSE2)
@@ -343,6 +345,11 @@
       else()
          list(APPEND _disable_vector_unit_list "fma4")
       endif()
+      if(USE_POPCNT)
+         list(APPEND _enable_vector_unit_list "popcnt")
+      else()
+         list(APPEND _disable_vector_unit_list "popcnt")
+      endif()
       if(MSVC)
          # MSVC on 32 bit can select /arch:SSE2 (since 2010 also /arch:AVX)
          # MSVC on 64 bit cannot select anything (should have changed with MSVC 2010)
@@ -440,6 +447,8 @@
                   set(_header "x86intrin.h")
                elseif(_flag STREQUAL "xop")
                   set(_header "x86intrin.h")
+               elseif(_flag STREQUAL "popcnt")
+                  set(_header "popcntintrin.h")
                endif()
                set(_resultVar "HAVE_${_header}")
                string(REPLACE "." "_" _resultVar "${_resultVar}")
diff -urP /var/tmp/portage/sci-libs/Vc-0.7.1-r2/work/Vc-0.7.1/CMakeLists.txt /tmp/Vc-0.7.1/CMakeLists.txt
--- /var/tmp/portage/sci-libs/Vc-0.7.1-r2/work/Vc-0.7.1/CMakeLists.txt	2014-02-20 19:58:45.000000000 +0100
+++ /tmp/Vc-0.7.1/CMakeLists.txt	2014-02-20 19:52:31.000000000 +0100
@@ -21,8 +21,12 @@
 SET(_REPLACE_PATTERNS "-march=[a-zA-Z0-9]* " "-msse " "-msse2 " "-msse3 " "-msse4 " "-msse4_1 " "-msse4_2 " "-avx " "-xop ")
 FOREACH(_PATTERN ${_REPLACE_PATTERNS})
        MESSAGE("Removing ${_PATTERN} from provided C(XX)_FLAGS")
-       STRING(REGEX REPLACE ${_PATTERN} "" CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
-       STRING(REGEX REPLACE ${_PATTERN} "" CMAKE_C_FLAGS ${CMAKE_C_FLAGS})
+       IF(NOT CMAKE_CXX_FLAGS STREQUAL "")
+              STRING(REGEX REPLACE ${_PATTERN} "" CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
+       ENDIF()
+       IF(NOT CMAKE_C_FLAGS STREQUAL "")
+              STRING(REGEX REPLACE ${_PATTERN} "" CMAKE_C_FLAGS ${CMAKE_C_FLAGS})
+       ENDIF()
 ENDFOREACH(_PATTERN)
 
 
diff -urP /var/tmp/portage/sci-libs/Vc-0.7.1-r2/work/Vc-0.7.1/avx/intrinsics.h /tmp/Vc-0.7.1/avx/intrinsics.h
--- /var/tmp/portage/sci-libs/Vc-0.7.1-r2/work/Vc-0.7.1/avx/intrinsics.h	2013-03-26 10:45:14.000000000 +0100
+++ /tmp/Vc-0.7.1/avx/intrinsics.h	2014-02-20 19:52:31.000000000 +0100
@@ -29,7 +29,7 @@
 // AVX
 #include <immintrin.h>
 
-#if (defined(VC_IMPL_XOP) || defined(VC_IMPL_FMA4)) && !defined(VC_MSVC)
+#if (defined(VC_IMPL_XOP) || defined(VC_IMPL_FMA4) || defined(VC_IMPL_POPCNT)) && !defined(VC_MSVC)
 #include <x86intrin.h>
 #endif
 }
diff -urP /var/tmp/portage/sci-libs/Vc-0.7.1-r2/work/Vc-0.7.1/sse/intrinsics.h /tmp/Vc-0.7.1/sse/intrinsics.h
--- /var/tmp/portage/sci-libs/Vc-0.7.1-r2/work/Vc-0.7.1/sse/intrinsics.h	2014-02-20 19:58:44.000000000 +0100
+++ /tmp/Vc-0.7.1/sse/intrinsics.h	2014-02-20 19:52:31.000000000 +0100
@@ -28,6 +28,10 @@
 // intrin.h header (included indirectly via unistd.h) declares many SSE intrinsics again.
 extern "C" {
 #include <immintrin.h>
+
+#if (defined(VC_IMPL_POPCNT)) && !defined(VC_MSVC)
+#include <x86intrin.h>
+#endif
 }
 
 #if defined(__GNUC__) && !defined(VC_IMPL_SSE2)
